### PR TITLE
Update test infrastructure and babel

### PR DESCRIPTION
**Pull request by adjerbetian**
I wanted to work on https://github.com/benmosher/eslint-plugin-import/issues/1481 **but the tests were failing locally**. This PR fixes that.

## Main change : babel configuration

The tests failed on master because the `require` function failed to load some package.

<details>

<summary>Error details</summary>

```
> cross-env BABEL_ENV=test NODE_PATH=./src nyc -s mocha -R dot --recursive -t 5s "tests/src"

internal/modules/cjs/loader.js:985
  throw err;
  ^

Error: Cannot find module 'core/importType'
Require stack:
- /home/djerbetian/code/eslint-plugin-import/tests/src/core/importType.js
- /home/djerbetian/code/eslint-plugin-import/node_modules/mocha/lib/mocha.js
- /home/djerbetian/code/eslint-plugin-import/node_modules/mocha/index.js
- /home/djerbetian/code/eslint-plugin-import/node_modules/mocha/bin/_mocha
    at Function.Module._resolveFilename (internal/modules/cjs/loader.js:982:15)
    at Function.Module._load (internal/modules/cjs/loader.js:864:27)
    at Module.require (internal/modules/cjs/loader.js:1044:19)
    at require (internal/modules/cjs/helpers.js:77:18)
```

</details>

The problem was from some dynamic require. I  fixed the problem here: https://github.com/adjerbetian/eslint-plugin-import/commit/3e592e68edc551a2ef8065066470e17d7b695797.

## Other change: mocha configuration

The current test configuration was spread between nyc ([here](https://github.com/benmosher/eslint-plugin-import/blob/1a3a12869651bec54d567af8dc0e7cb2a801d41f/package.json#L107)) and [mocha](https://github.com/benmosher/eslint-plugin-import/blob/1a3a12869651bec54d567af8dc0e7cb2a801d41f/package.json#L31). I migrated this babel compiler configuration to `.mocharc` so it's easily reusable. See here: https://github.com/adjerbetian/eslint-plugin-import/commit/633bfeb81556600693b74a1e79c6716c09df2b1c.

*Note*: as I updated mocha so it supports `.mocharc`, it removed the `lodash.isarray` dependency that was assumed to be present in some tests, so I added it manually.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -53,15 +53,19 @@
   },
   "homepage": "https://github.com/benmosher/eslint-plugin-import",
   "devDependencies": {
+    "@babel/cli": "^7.8.4",
+    "@babel/core": "^7.8.7",
+    "@babel/preset-env": "^7.8.7",
+    "@babel/register": "^7.8.6",
     "@eslint/import-test-order-redirect-scoped": "file:./tests/files/order-redirect-scoped",
     "@test-scope/some-module": "file:./tests/files/symlinked-module",
+    "@types/chai": "^4.2.11",
```
Comment (ljharb) [inline package.json]: This project isn’t in typescript, so these types aren’t useful.

File: resolvers/webpack/.babelrc.js [hunk 2]
```diff
@@ -0,0 +1 @@
+../../.babelrc.js
```
Comment (ljharb) [inline resolvers/webpack/.babelrc.js]: I prefer non-js configs only if possible.

**Comment by adjerbetian** [comment]
@ljharb Thank you for your answers!

Ok, I'll try to make the tests pass without touching mocha and babel.

Thanks again !

File: .gitignore [hunk 3]
```diff
@@ -49,3 +49,6 @@ lib/
 yarn.lock
 package-lock.json
 npm-shrinkwrap.json
+
+# IDE
+.idea
```
Comment (ljharb) [inline .gitignore]: this can go in your global gitignore rather than in every repo you touch :-)

File: .mocharc.yml [hunk 4]
```diff
@@ -0,0 +1,5 @@
+recursive: true
```
Comment (ljharb) [inline .mocharc.yml]: I prefer the “.mocharc” filename.

File: .gitignore (hunk 3 above)
Comment (adjerbetian) [reply .gitignore]: @ljharb True, but on some projects, I like to commit some [IDE configuration](https://intellij-support.jetbrains.com/hc/en-us/community/posts/207656425-Arguments-for-NOT-adding-the-entire-idea-folder-to-gitignore-hgignore). It makes it easier also to share some configuration with other people, and with myself when I switch computers :sweat_smile:.

File: package.json (hunk 1 above)
Comment (adjerbetian) [reply package.json]: @ljharb ok, I removed them. :+1:

Usually, I still like to have them still as it improves the IDE autocompletion. But there are ways around it.

File: resolvers/webpack/.babelrc.js (hunk 2 above)
Comment (adjerbetian) [reply resolvers/webpack/.babelrc.js]: @ljharb I removed this file and the `mocharc` file, so I resolve this conversation, so I allow myself to resolve here.

**Comment by adjerbetian** [comment]
@ljharb I simplified my modification to only make my local tests pass. What do you think?

File: .gitignore (hunk 3 above)
Comment (ljharb) [reply .gitignore]: That can be done via dotfiles repos and similar :-)

File: .babelrc [hunk 5]
```diff
@@ -2,8 +2,16 @@
   "presets": [ "es2015-argon" ],
   "sourceMaps": "inline",
   "env": {
-    "test": {
-      "plugins": [ "istanbul" ]
+    "src": {
```
Comment (ljharb) [inline .babelrc]: generally the "env" is test/prod/dev, not a file location; this suggests that perhaps the production files would ship with the istanbul transform?

perhaps naming these envs "test" and "test-compiled" would be clearer :-)

File: .gitignore (hunk 3 above)
Comment (ljharb) [inline .gitignore]:
```suggestion
```

File: package.json [hunk 6]
```diff
@@ -25,10 +25,10 @@
     "watch": "npm run mocha -- --watch tests/src",
     "pretest": "linklocal",
     "posttest": "eslint .",
-    "mocha": "cross-env BABEL_ENV=test NODE_PATH=./src nyc -s mocha -R dot --recursive -t 5s",
+    "mocha": "cross-env BABEL_ENV=src nyc -s mocha -R dot --recursive -t 5s",
     "tests-only": "npm run mocha tests/src",
     "test": "npm run tests-only",
-    "test-compiled": "npm run prepublish && NODE_PATH=./lib mocha --compilers js:babel-register --recursive tests/src",
+    "test-compiled": "npm run prepublish && BABEL_ENV=lib mocha --compilers js:babel-register --recursive tests/src",
```
Comment (ljharb) [inline package.json]: i *very much* like avoiding use of the terrible and horrific `NODE_PATH`, and using babel for this instead

File: tests/src/package.js [hunk 7]
```diff
@@ -51,10 +51,17 @@ describe('package', function () {
...
+        expect(() => require(getRulePath(rule.slice(preamble.length))))
           .not.to.throw(Error)
       }
     }
+
+    function getRulePath(ruleName) {
+      // 'require' does not work with dynamic paths because of the compilation step by babel
+      // (which resolves paths according to the root folder configuration)
+      // the usage of require.resolve on a static path gets around this
+      return path.resolve(require.resolve('rules/no-unresolved'), '../', ruleName)
```
Comment (ljharb) [inline tests/src/package.js]:
```suggestion
      return path.resolve(require.resolve('rules/no-unresolved'), '..', ruleName)
```

File: .gitignore (hunk 3 above)
Comment (adjerbetian) [reply .gitignore]: @ljharb I'm not sure of what you mean by "*dotfiles repo*", but if you mean a configuration repository common to all projects (and team), that is a complementary option, but some configurations are still custom to the project and therefore belong in the project itself. I see you even have a file like this here: [`import.sublime-project`](https://github.com/benmosher/eslint-plugin-import/blob/master/import.sublime-project).

But ok, I understand you really don't want to have this added to the `.gitignore`, I don't mind :wink:. I removed it here: https://github.com/adjerbetian/eslint-plugin-import/commit/ba0d05ad1ae5f3e77ed741782a23031713221925

File: tests/src/package.js (hunk 7 above)
Comment (adjerbetian) [reply tests/src/package.js]: @ljharb :+1: Corrected here: https://github.com/adjerbetian/eslint-plugin-import/commit/ee23f8fad303eac93da3de8d767d4775ea3a8d28

File: .babelrc (hunk 5 above)
Comment (adjerbetian) [reply .babelrc]: @ljharb Yes very true. Unfortunately `test-compiled` didn't work, it looks like I could not put a dash in the env name. I used `testCompiled` instead: https://github.com/adjerbetian/eslint-plugin-import/commit/a6ed0b3c33e03e6fd0562dbc5c0cc2d447fb5f05

**Comment by adjerbetian** [comment]
@ljharb My first open source contribution! :tada::tada::tada: As small as it is, I still celebrate :champagne:

PS: if you have remarks on how I worked, please to. For instance, I'm not sure whether I should have pingued you when answering your comments.

**Comment by ljharb** [comment]
Everything was fine :-) some people may only respond when they get a direct ping, perhaps if they've unsubscribed, so it's probably best to save explicit mentions for when it seems like the person will never otherwise see it.

**Pull request merged**